src/main/java/com/example/auctionapp/seeder/DatabaseSeeder.java

```diff
@@ -23,25 +25,35 @@
 
     private BaseRepository<Role> roleRepository;
     private UserRepository userRepository;
+    private BaseRepository<Category> categoryRepository;
+    private BaseRepository<Subcategory> subcategoryRepository;
 
     private Logger logger = LoggerFactory.getLogger(RoleService.class);
     private PasswordEncoder passwordEncoder;
 
     @Autowired
     public DatabaseSeeder(BaseRepository<Role> roleRepository,
                           UserRepository userRepository,
-                          PasswordEncoder passwordEncoder) {
+                          PasswordEncoder passwordEncoder,
+                          BaseRepository<Category> categoryRepository,
+                          BaseRepository<Subcategory> subcategoryRepository) {
         this.roleRepository = roleRepository;
         this.roleRepository.setResourceClass(Role.class);
         this.userRepository = userRepository;
         this.userRepository.setResourceClass(User.class);
         this.passwordEncoder = passwordEncoder;
+        this.categoryRepository = categoryRepository;
+        this.categoryRepository.setResourceClass(Category.class);
+        this.subcategoryRepository=subcategoryRepository;
+        this.subcategoryRepository.setResourceClass(Subcategory.class);
```

Comment:
I can see this becoming a problem in the long run.
Somehow it should be given that if a generic class is created with a specific type for `T`, it would be posible to access that specific generic type from inside the class.
This is done via reflection, or you can create a constructor in BaseRepository that accepts the specific type T and create beans instead of letting spring autowire them.
The reflection method is done like this:
```
type = ((Class) ((ParameterizedType) getClass()
      .getGenericSuperclass()).getActualTypeArguments()[0])
```
this should be put in the constructor.
Either way, we need to create a constructor for `BaseRepository.class`

Reply:
Reflection is an advance java technique, we wont use it in the scope of this up except for this, but you can feel free to read about it

Reply:
And one more thing, prefer using `constructors` and avoid creating setters for classes that are not models. Those classes are usually services etc, and their properties should be set specifically on creation - via the constructors - never to be set later again. (`private final`)
`private` - so nothing can access them except the class (or `protected` if needed)
`final` - nothing can change the instance after it has been set.